crates/red_knot_python_semantic/src/semantic_index/builder/except_handlers.rs

```diff
@@ -0,0 +1,128 @@
+use std::cell::{RefCell, RefMut};
+
+use crate::semantic_index::use_def::FlowSnapshot;
+
+use super::SemanticIndexBuilder;
+
+/// An abstraction over the fact that each scope should have its own [`TryNodeContextStack`]
+#[derive(Debug, Default)]
+pub(super) struct TryNodeContextStackManager(Vec<TryNodeContextStack>);
+
+impl TryNodeContextStackManager {
+    /// Push a new [`TryNodeContextStack`] onto the stack of stacks.
+    ///
+    /// Each [`TryNodeContextStack`] is only valid for a single scope
+    pub(super) fn enter_nested_scope(&mut self) {
+        self.0.push(TryNodeContextStack::default());
+    }
+
+    /// Retrieve the [`TryNodeContextStack`] that is relevant for the current scope.
+    pub(super) fn current_try_context_stack(&self) -> &TryNodeContextStack {
+        self.0
+            .last()
+            .expect("There should always be at least one `TryBlockContexts` on the stack")
+    }
+
+    /// Pop a new [`TryNodeContextStack`] off the stack of stacks.
+    ///
+    /// Each [`TryNodeContextStack`] is only valid for a single scope
+    pub(super) fn exit_scope(&mut self) {
+        let popped_context = self.0.pop();
+        assert!(
+            popped_context.is_some(),
+            "pop_context() should never be called on an empty stack \
```

Comment:
```suggestion
            "exit_scope() should never be called on an empty stack \
```

Nit: I think it would also be fine to make this a debug assertion.